src/node_osrm.cpp

```diff
@@ -600,46 +513,53 @@ void Engine::Run(_NAN_METHOD_ARGS, route_parameters_ptr params)
     closure->machine = ObjectWrap::Unwrap<Engine>(args.This());
     closure->params = std::move(params);
     closure->error = "";
-    NanAssignPersistent(closure->cb, callback.As<Function>());
-    uv_queue_work(uv_default_loop(), &closure->request, AsyncRun, reinterpret_cast<uv_after_work_cb>(AfterRun));
+    closure->cb.Reset(callback.As<v8::Function>());
+    uv_queue_work(uv_default_loop(), &closure->request, AsyncRun,
+                  reinterpret_cast<uv_after_work_cb>(AfterRun));
     closure->machine->Ref();
     return;
 }
 
-void Engine::AsyncRun(uv_work_t* req) {
+void Engine::AsyncRun(uv_work_t *req)
+{
     RunQueryBaton *closure = static_cast<RunQueryBaton *>(req->data);
-    try {
+    try
+    {
         closure->machine->this_->RunQuery(*closure->params, closure->result);
-    } catch(std::exception const& ex) {
+    }
+    catch (std::exception const &ex)
+    {
         closure->error = ex.what();
     }
 }
 
-void Engine::AfterRun(uv_work_t* req) {
-    NanScope();
+void Engine::AfterRun(uv_work_t *req)
+{
     RunQueryBaton *closure = static_cast<RunQueryBaton *>(req->data);
-    TryCatch try_catch;
-    if (closure->error.size() > 0) {
-        Local<Value> argv[1] = { NanError(closure->error.c_str()) };
-        NanMakeCallback(NanGetCurrentContext()->Global(), NanNew(closure->cb), 1, argv);
-    } else {
-        Local<Value> result;
+    Nan::TryCatch try_catch;
+    if (closure->error.size() > 0)
+    {
+        v8::Local<v8::Value> argv[1] = {Nan::Error(closure->error.c_str())};
+        Nan::MakeCallback(Nan::GetCurrentContext()->Global(), Nan::New(closure->cb), 1, argv);
+    }
+    else
+    {
+        v8::Local<v8::Value> result;
         osrm::json::render(result, closure->result);
-        Local<Value> argv[2] = { NanNull(), result };
-        NanMakeCallback(NanGetCurrentContext()->Global(), NanNew(closure->cb), 2, argv);
+        v8::Local<v8::Value> argv[2] = {Nan::Null(), result};
+        Nan::MakeCallback(Nan::GetCurrentContext()->Global(), Nan::New(closure->cb), 2, argv);
     }
-    if (try_catch.HasCaught()) {
-        node::FatalException(try_catch);
+    if (try_catch.HasCaught())
+    {
+        Nan::FatalException(try_catch);
     }
     closure->machine->Unref();
-    NanDisposePersistent(closure->cb);
+    closure->cb.Reset();
     delete closure;
```

Comment:
Sure, but the cleanup should still happen in a RAII-way, that is at scope exit.

Boost.ScopeExit let's you defer actions, otherwise:

    template <typename Fn>
    struct ScopeExit {
      ScopeExit(Fn fn_) : fn{fn_} {}
      ~ScopeExit{ fn(); }
      Fn fn;
    };